result_iterator.go

```diff
@@ -48,26 +48,32 @@ func (r *queryResultIterator) More() bool {
 		return false
 	}
 
-	if r.results == nil {
-		results, ok := <-r.query.Ready()
-		if !ok {
-			return false
-		}
-		r.results = NewMapResultIterator(results)
+	if r.nextResult != nil {
+		return true
 	}
-	return r.results.More()
+
+	nr, ok := <-r.query.Ready()
+	if !ok {
+		return false
+	}
+
+	r.nextResult = nr
+	return true
 }
 
 func (r *queryResultIterator) Next() Result {
-	return r.results.Next()
+	if r.nextResult == nil {
+		r.nextResult = <-r.query.Ready()
+	}
```

Comment:
Next is allowed to panic if More returns `false` and `Next` is called. That means that this check is not necessary and could possibly lead to odd bugs.

Reply:
I don't think that would quite get the behavior we want, but you're right there's logic errors here. I added a new commit (with comments) that addresses them.